File: gamfit/_fidelity_metrics.py
"""Model-currency fidelity metric reductions (scorecard axis 1).

Thin wrappers over the Rust source of truth ``gam_math::fidelity_metrics``
(exposed through ``gamfit._rust`` as the ``fidelity_*`` FFI): ``loss_recovered``,
Euclidean ``r2_score``, categorical ``kl_categorical_rows`` over token rows, and
the ``distortion_floor_r2``. Held-out EV prices reconstruction error in Euclidean
distance, but the model reads directions unequally; these price reconstruction in
the model's own loss/output currency (see METRICS.md axis 1). The *patching* that
produces the losses/logprobs lives at the torch/HF boundary in the harness; these
are the pure reductions, shared verbatim by every consumer (CONTROL composed-dict
run, EVAL scorecard, RUNG1 iid-vs-Fisher-GLS A/B) so the program has ONE currency.

Every function routes unconditionally to the Rust kernels — there is no numpy
twin. numpy here is marshalling only (shape validation, flattening to the
row-major flat layout the FFI consumes). If the compiled extension is missing the
``fidelity_*`` entry points that is an ImportError bug in the build, not a reason
to reimplement the math.

Definitions (model loss units / nats):
    loss_recovered = (L_ablate − L_recon) / (L_ablate − L_clean)
    r2_score       = 1 − Σ(clean − approx)² / Σ(clean − colmean(clean))²
    kl_categorical_rows = mean_row Σ_v p·(logp_clean − logp_other), p = exp(logp_clean)
    distortion_floor_r2  = highest quant-R² at which loss_recovered has fallen
                           more than ``tol_frac`` below the finest-precision plateau
"""

from __future__ import annotations

from typing import Optional, Sequence

import numpy as np


def _rust_metrics():
    """Return the compiled extension, requiring the ``fidelity_*`` FFI kernels.

    Hard dependency: the metric math is owned by Rust. A missing extension or a
    wheel that predates the ``fidelity_*`` exposure is a build bug and surfaces as
    an ImportError rather than a silent numpy reimplementation.
    """
    from gamfit._binding import rust_module

    module = rust_module()
    if not hasattr(module, "fidelity_loss_recovered"):
        raise ImportError(
            "gamfit._rust is missing the fidelity_* FFI kernels; rebuild the wheel "
            "(gam_math::fidelity_metrics must be exposed through gam-pyffi)"
        )
    return module


def loss_recovered(l_clean: float, l_recon: float, l_ablate: float) -> float:
    """(L_ablate − L_recon) / (L_ablate − L_clean); NaN when the gap is degenerate."""
    return float(
        _rust_metrics().fidelity_loss_recovered(float(l_clean), float(l_recon), float(l_ablate))
    )


def r2_score(clean: np.ndarray, approx: np.ndarray) -> float:
    """Euclidean R² of ``approx`` vs ``clean``; TSS about the per-column mean of clean."""
    clean = np.ascontiguousarray(clean, dtype=np.float64)
    approx = np.ascontiguousarray(approx, dtype=np.float64)
    if clean.shape != approx.shape:
        raise ValueError(f"r2_score shape mismatch: {clean.shape} vs {approx.shape}")
    clean2 = clean.reshape(clean.shape[0], -1) if clean.ndim > 1 else clean.reshape(1, -1)
    approx2 = approx.reshape(clean2.shape)
    n_rows, n_cols = clean2.shape
    return float(
        _rust_metrics().fidelity_r2_score(
            clean2.ravel().tolist(), approx2.ravel().tolist(), int(n_rows), int(n_cols)
        )
    )


def kl_categorical_rows(clean_logprobs: np.ndarray, other_logprobs: np.ndarray) -> float:
    """Mean over rows of KL(clean ‖ other) from natural-log probability rows (nats)."""
    clean_lp = np.ascontiguousarray(clean_logprobs, dtype=np.float64)
    other_lp = np.ascontiguousarray(other_logprobs, dtype=np.float64)
    if clean_lp.shape != other_lp.shape:
        raise ValueError(f"kl shape mismatch: {clean_lp.shape} vs {other_lp.shape}")
    clean2 = clean_lp.reshape(clean_lp.shape[0], -1) if clean_lp.ndim > 1 else clean_lp.reshape(1, -1)
    other2 = other_lp.reshape(clean2.shape)
    n_rows, n_cols = clean2.shape
    return float(
        _rust_metrics().fidelity_kl_categorical_rows(
            clean2.ravel().tolist(), other2.ravel().tolist(), int(n_rows), int(n_cols)
        )
    )


def distortion_floor_r2(
    r2s: Sequence[float], loss_recovereds: Sequence[float], tol_frac: float = 0.05
) -> Optional[float]:
    """Highest-R² point at which loss_recovered drops >``tol_frac`` below the plateau.

    ``r2s`` and ``loss_recovereds`` are parallel per-quantisation-level arrays
    (order irrelevant; sorted by R² descending internally). Above the returned R²
    extra precision buys ~no model-usable fidelity, so all fidelity is read AT the
    floor. Returns ``None`` only for an empty sweep.
    """
    r2_arr = np.asarray(r2s, dtype=np.float64)
    lr_arr = np.asarray(loss_recovereds, dtype=np.float64)
    if r2_arr.shape != lr_arr.shape:
        raise ValueError("distortion_floor_r2: parallel arrays length mismatch")
    out = _rust_metrics().fidelity_distortion_floor_r2(
        r2_arr.tolist(), lr_arr.tolist(), float(tol_frac)
    )
    return None if out is None else float(out)

File: crates/gam-math/src/fidelity_metrics.rs
//! Model-currency fidelity metrics (scorecard axis 1).
//!
//! A featurizer's held-out EV prices reconstruction error in Euclidean distance,
//! but the model reads directions unequally (rogue dims: huge variance, ~0
//! leverage). These reductions price reconstruction in the MODEL's own currency,
//! by comparing losses (and output distributions) when the harvest-layer
//! activation is run clean, replaced by the reconstruction, or replaced by the
//! information-free mean-ablation baseline. They are the reusable numeric core
//! behind the Python fidelity harness (patching itself lives at the torch/HF
//! boundary); the harness's pure-Python fallback is pinned to these functions to
//! double precision by `tests/metrics/test_fidelity_metrics_parity.py`.
//!
//! Definitions (all in the model's loss units / nats):
//!   loss_recovered = (L_ablate − L_recon) / (L_ablate − L_clean)
//!     1 = reconstruction recovers the clean loss; 0 = no better than ablation.
//!   r2_score       = 1 − Σ(clean − approx)² / Σ(clean − colmean(clean))²
//!     Euclidean explained variance about the per-column mean of `clean`.
//!   kl_categorical_rows = mean_row Σ_v p·(logp_clean − logp_other), p = exp(logp_clean)
//!     mean over rows of KL(clean ‖ other) from natural-log probabilities (nats).
//!   distortion_floor_r2 = the highest quantisation-R² at which loss_recovered
//!     has fallen more than `tol_frac` below the finest-precision plateau — the
//!     precision floor below which model-usable fidelity starts to degrade, and
//!     above which extra precision is wasted (read all metrics AT the floor).

/// (L_ablate − L_recon) / (L_ablate − L_clean).
///
/// Returns NaN when the clean↔ablate loss gap is degenerate (|denom| < `EPS`),
/// so a flat baseline surfaces as not-a-number rather than a divide blow-up.
pub fn loss_recovered(l_clean: f64, l_recon: f64, l_ablate: f64) -> f64 {
    const EPS: f64 = 1e-12;
    let denom = l_ablate - l_clean;
    if denom.abs() < EPS {
        return f64::NAN;
    }
    (l_ablate - l_recon) / denom
}

/// Euclidean R² of `approx` against `clean`, both row-major `(n_rows, n_cols)`.
///
/// TSS is taken about the per-column mean of `clean` (the same convention the
/// harness's `r2_against` uses). Returns 0.0 when TSS is non-positive (a constant
/// `clean`), matching the reference's guard.
pub fn r2_score(clean: &[f64], approx: &[f64], n_rows: usize, n_cols: usize) -> f64 {
    assert_eq!(clean.len(), n_rows * n_cols, "clean shape mismatch");
    assert_eq!(approx.len(), n_rows * n_cols, "approx shape mismatch");
    if n_rows == 0 || n_cols == 0 {
        return 0.0;
    }
    let mut col_mean = vec![0.0_f64; n_cols];
    for row in 0..n_rows {
        let base = row * n_cols;
        for col in 0..n_cols {
            col_mean[col] += clean[base + col];
        }
    }
    let inv_rows = 1.0 / n_rows as f64;
    for value in col_mean.iter_mut() {
        *value *= inv_rows;
    }
    let mut rss = 0.0_f64;
    let mut tss = 0.0_f64;
    for row in 0..n_rows {
        let base = row * n_cols;
        for col in 0..n_cols {
            let c = clean[base + col];
            let residual = c - approx[base + col];
            rss += residual * residual;
            let centered = c - col_mean[col];
            tss += centered * centered;
        }
    }
    if tss > 0.0 { 1.0 - rss / tss } else { 0.0 }
}

/// Mean over rows of KL(clean ‖ other) from natural-log probability rows.
///
/// Each of `clean_logprobs`/`other_logprobs` is row-major `(n_rows, n_cols)` of
/// log-probabilities (natural log). Per row: Σ_v exp(logp_clean)·(logp_clean −
/// logp_other); averaged over rows. Units are nats. `n_cols` is the vocabulary
/// width.
pub fn kl_categorical_rows(
    clean_logprobs: &[f64],
    other_logprobs: &[f64],
    n_rows: usize,
    n_cols: usize,
) -> f64 {
    assert_eq!(
        clean_logprobs.len(),
        n_rows * n_cols,
        "clean logprobs shape mismatch"
    );
    assert_eq!(
        other_logprobs.len(),
        n_rows * n_cols,
        "other logprobs shape mismatch"
    );
    if n_rows == 0 {
        return 0.0;
    }
    let mut total = 0.0_f64;
    for row in 0..n_rows {
        let base = row * n_cols;
        let mut row_kl = 0.0_f64;
        for col in 0..n_cols {
            let logp_clean = clean_logprobs[base + col];
            let p = logp_clean.exp();
            row_kl += p * (logp_clean - other_logprobs[base + col]);
        }
        total += row_kl;
    }
    total / n_rows as f64
}

/// The distortion-floor R²: the highest-R² point at which `loss_recovered` has
/// dropped more than `tol_frac` (relative) below the finest-precision plateau.
///
/// `r2s` and `loss_recovered` are parallel arrays over quantisation levels (one
/// entry per level; order irrelevant — the routine sorts by R² descending). The
/// plateau is the loss_recovered at the highest R² (finest precision); the floor
/// is the first (in descending-R² order) point that falls below `plateau −
/// tol_frac·|plateau|`. If nothing drops, the coarsest R² is returned. `None`
/// only for an empty sweep.
pub fn distortion_floor_r2(r2s: &[f64], loss_recovered: &[f64], tol_frac: f64) -> Option<f64> {
    assert_eq!(
        r2s.len(),
        loss_recovered.len(),
        "parallel arrays length mismatch"
    );
    if r2s.is_empty() {
        return None;
    }
    let mut order: Vec<usize> = (0..r2s.len()).collect();
    // Descending by R² (finest precision first). Total order via partial_cmp with
    // a stable tie-break on index so the plateau pick is deterministic.
    order.sort_by(|&a, &b| {
        r2s[b]
            .partial_cmp(&r2s[a])
            .unwrap_or(std::cmp::Ordering::Equal)
            .then(a.cmp(&b))
    });
    let plateau = loss_recovered[order[0]];
    let threshold = plateau - tol_frac * plateau.abs();
    for &idx in &order {
        if loss_recovered[idx] < threshold {
            return Some(r2s[idx]);
        }
    }
    Some(r2s[order[order.len() - 1]])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loss_recovered_endpoints_and_degenerate() {
        // recon == clean -> full recovery = 1; recon == ablate -> 0.
        assert!((loss_recovered(1.0, 1.0, 3.0) - 1.0).abs() < 1e-15);
        assert!((loss_recovered(1.0, 3.0, 3.0) - 0.0).abs() < 1e-15);
        // halfway
        assert!((loss_recovered(1.0, 2.0, 3.0) - 0.5).abs() < 1e-15);
        // degenerate gap -> NaN
        assert!(loss_recovered(2.0, 2.0, 2.0).is_nan());
    }

    #[test]
    fn r2_perfect_and_mean_baselines() {
        let clean = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0];
        // perfect reconstruction -> 1
        assert!((r2_score(&clean, &clean, 3, 2) - 1.0).abs() < 1e-15);
        // predict the per-column mean everywhere -> R² = 0
        // col means: col0 mean of {1,3,5}=3, col1 mean of {2,4,6}=4
        let mean_pred = [3.0, 4.0, 3.0, 4.0, 3.0, 4.0];
        assert!(r2_score(&clean, &mean_pred, 3, 2).abs() < 1e-15);
    }

    #[test]
    fn r2_constant_clean_is_zero() {
        let clean = [7.0, 7.0, 7.0, 7.0];
        let approx = [1.0, 2.0, 3.0, 4.0];
        assert_eq!(r2_score(&clean, &approx, 2, 2), 0.0);
    }

    #[test]
    fn kl_identical_is_zero_and_positive_otherwise() {
        // two rows, 3-way categorical, natural-log probs
        let p = [0.5_f64, 0.25, 0.25, 0.1, 0.6, 0.3];
        let logp: Vec<f64> = p.iter().map(|v| v.ln()).collect();
        assert!(kl_categorical_rows(&logp, &logp, 2, 3).abs() < 1e-15);
        // uniform "other" -> KL = mean row (Σ p logp + log 3) > 0
        let uniform = vec![(1.0_f64 / 3.0).ln(); 6];
        let kl = kl_categorical_rows(&logp, &uniform, 2, 3);
        // closed form: mean over rows of (-H(p) + log 3)
        let mut expect = 0.0;
        for row in 0..2 {
            let mut h = 0.0;
            for col in 0..3 {
                let pv = p[row * 3 + col];
                h += pv * pv.ln();
            }
            expect += h + (3.0_f64).ln();
        }
        expect /= 2.0;
        assert!((kl - expect).abs() < 1e-15);
    }

    #[test]
    fn floor_flat_plateau_returns_coarsest() {
        // loss_recovered flat across all precisions -> nothing drops -> coarsest R².
        let r2s = [0.99, 0.90, 0.50];
        let lr = [1.0, 1.0, 1.0];
        assert_eq!(distortion_floor_r2(&r2s, &lr, 0.05), Some(0.50));
    }

    #[test]
    fn floor_detects_drop_point() {
        // finest plateau lr=1.0; coarse point drops to 0.8 (>5% below) at R²=0.60.
        let r2s = [0.99, 0.95, 0.60, 0.30];
        let lr = [1.0, 0.99, 0.80, 0.40];
        // descending R²: 0.99(1.0 plateau), 0.95(0.99 ok), 0.60(0.80 < 0.95) -> floor 0.60
        assert_eq!(distortion_floor_r2(&r2s, &lr, 0.05), Some(0.60));
    }

    #[test]
    fn floor_empty_is_none() {
        assert_eq!(distortion_floor_r2(&[], &[], 0.05), None);
    }
}
